### PR TITLE
Check for duplicate NuGet items, and raise warnings/errors where appropriate

### LINKED ISSUE #9864: Duplicate PackageReference includes are handled inconsistently by all restores, leading to unnecessary full restores
Take a project such as the following: 

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NuGet.Protocol" Version="5.5.0" />
    <PackageReference Include="NuGet.Protocol" Version="5.6.0" />
  </ItemGroup>

</Project>
```

Note the duplicate PackageReference declaration. Digging further reveals that different project types and different tools handle it differently. 

See the below table for details

| Restore Flavor | PackageReference declaration | 
| - | - |
| Commandline restore (all) | First |
| Static graph restore (all) | First | 
| VS - SDK based projects | Last |
| VS - Legacy PR projects | First |

This leads to issues as the ones described in https://github.com/dotnet/arcade/issues/5550 (see below repro). 

I think logically from an msbuild perspective, we'd expect that the last value is the one that's resolved. This needs investigated and ideally consolidated. 
Note that consolidating VS - LEgacy PR projects might be impossible, due to the fact that it's largely in maintenance mode.

See repro below.

cc folks that might have some thoughts around: 

@jeffkl @davkean @dsplaisted @zivkan 

fyi @sharwell @garath

[DuplicateIds.zip](https://github.com/NuGet/Home/files/5002828/DuplicateIds.zip)

**WIP**: 
**Design**: https://github.com/NuGet/Home/compare/dev-nkolev92-duplicatesHandling 
**Implementation**: https://github.com/NuGet/NuGet.Client/tree/dev-nkolev92-duplicatePRs

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9864

Regression? Last working version:

## Description

Refer to design in https://github.com/NuGet/Home/pull/11649 for a detailed description.

When a NuGet item is seen as a duplicate, a warning is raised. 
This warning supports NoWarn, TreatWarningsAsErrors, and it can be disabled altogether. 

We're also supporting ContinueOnError for VS purposes.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled - https://github.com/NuGet/docs.microsoft.com-nuget/issues/2700
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by zivkan** [comment]
I can't remember what parts of NuGet.targets that static graph restore uses, and what parts it overwrites.  Do these new NU codes still work in static restore?

**Comment by nkolev92** [comment]
@zivkan 

static graph still calls the CollectXYZ targets, so it'll raise the warnings.

**Comment by nkolev92** [comment]
Note that I *might* change the warnings to NU1504/NU1505/NU1506 if/when https://github.com/NuGet/NuGet.Client/pull/4492 is merged.

File: src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs:18]: I think this task should implement    https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.framework.icancelabletask interface so that it supports cancellations.
I remember seeing an issue on NuGet/Home where an issue was created to add support for 'GetRestoreSettings' task. Please let me know if I'm missing anything.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs:18]: Maybe that issue can add cancellable interface to all the tasks?

There's a big difference between GetRestoreSettings and this task though.  
GetRestoreSettings does a lot of IO which can be slow, while this is just simple compute, so I don't think it'd do much.

Comment (zivkan) [reply src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs:18]: > Maybe that issue can add cancellable interface to all the tasks?

I'm not a fan of "existing code does it this way, so if we want to change the design patterns we use, do it in another issue where we change the whole repo".  Making it all-or-nothing raises the barrier so high that many things will never get implemented. So, I'm a strong believer in incremental improvements when modifying some code, but especially when writing new code.

> this is just simple compute, so I don't think it'd do much.

I agree. Looking at the code below, I can't see anywhere where it's reasonable/worthwhile to check a cancellation token.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs:18]: > existing code does it this way, so if we want to change the design patterns we use, do it in another issue where we change the whole repo

I think that's a reach. It's more of, on first look, this task makes no sense as ICancellable, so that other task can spend the time investigating and change it if they really think otherwise. 

Resolving the comment thread as this task is simply not suitable for adding cancellation.

**Comment by erdembayar** [comment]
LGTM. I can approve it after green build.

File: src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs:44]: Will there be any issues with `StringComparer.OrdinalIgnoreCase` option when 2 package references are same culturally but different when compared using the current approach?

For e.g. https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=net-6.0

//      encyclopædia = encyclopaedia (CurrentCultureIgnoreCase): True
//      encyclopædia = encyclopaedia (InvariantCultureIgnoreCase): True
//       encyclopædia = encyclopaedia (OrdinalIgnoreCase): False

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:182]: How about using `ContinueOnError` property directly instead of creating a new property? I see that `CollectPackageReferencesContinueOnError` is only used while invoking `CheckForDuplicateNuGetItemsTask`.

File: src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/CheckForDuplicateNuGetItemsTask.cs:44]: The package id comparer is OrdinalIgnoreCase all throughout the codebase. 
This is not a concern as https://github.com/NuGet/NuGet.Client/blob/2380e11075e50104862c499515a0d3e293a7774a/src/NuGet.Core/NuGet.Build.Tasks/GetRestorePackageReferencesTask.cs#L43 is already using that comparer.

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:182]: In some scenarios this target is called with no  value, and it basically leads to calling the Task with ContinueOnError empty which isn't a valid value: 

https://docs.microsoft.com/en-us/visualstudio/msbuild/task-element-msbuild?view=vs-2022

This solves that problem.

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:185]: What is the default value of `DisableCheckingDuplicateNuGetItems` property? If it is `false` then this task will execute because the condition evaluates to true.

**Review by kartheekp-ms — DISMISSED** [review]

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:185]: There's no default, so it's ''

**Review by erdembayar — DISMISSED** [review]

**Comment by nkolev92** [comment]
Thanks for the review @kartheekp-ms and @erdembayar 

Just waiting for dev to switch to 6.3 before merging this on early April.

**Comment by nkolev92** [comment]
Had to rebase to resolve merge conflicts. @kartheekp-ms, @erdembayar can one of you approve again?


Thanks!

**Review by erdembayar — APPROVED** [review]

**Comment by jaredpar** [comment]
FYI: this change caused a lot of frustration for internal customers updating to a new version of VS. The result is several teams are reverting the VS update that added it. 

Why was this diagonstic introduced in a minor update vs. a major one?

**Pull request merged**
